x-pack/plugins/monitoring/server/plugin.ts

```diff
@@ -0,0 +1,324 @@
+import { TelemetryCollectionManager } from 'src/legacy/core_plugins/telemetry/server/collection_manager';
+import { LOGGING_TAG, KIBANA_MONITORING_LOGGING_TAG } from '../common/constants';
+import {
+  Logger,
+  PluginInitializerContext,
+  RequestHandlerContext,
+  KibanaRequest,
+  KibanaResponseFactory,
+  CoreSetup,
+  ICustomClusterClient,
+  ElasticsearchConfig,
+  ElasticsearchConfigType,
+} from '../../../../src/core/server';
+import { MonitoringConfig } from './config';
+// @ts-ignore
+import { requireUIRoutes } from './routes';
+// @ts-ignore
+import { instantiateClient } from './es_client/instantiate_client';
+// @ts-ignore
+import { initBulkUploader, registerCollectors } from './kibana_monitoring';
+// @ts-ignore
+import { initInfraSource } from './lib/logs/init_infra_source';
+import { registerMonitoringCollection } from './telemetry_collection';
+import { XPackMainPlugin } from '../../../legacy/plugins/xpack_main/server/xpack_main';
+import { LicensingPluginSetup } from '../../licensing/server';
+import { PluginSetupContract } from '../../features/server';
+import { LicenseService } from './license_service';
+import { MonitoringLicenseService } from './types';
+
+export interface LegacyAPI {
+  xpackMain: XPackMainPlugin;
+  telemetryCollectionManager: TelemetryCollectionManager;
+  elasticsearch: any;
+  opsInterval: number;
+  getOSInfo: () => any;
+  hapiServer: any;
+  getServerStatus: () => string;
+  serverEvents: any;
+  infra: any;
+}
+
+interface PluginsSetup {
+  usageCollection: UsageCollectionSetup;
+  licensing: LicensingPluginSetup;
+  features: PluginSetupContract;
+}
+
+interface MonitoringCoreConfig {
+  get: (key: string) => string;
+}
+interface MonitoringCore {
+  config?: () => MonitoringCoreConfig;
+  log?: Logger;
+  route?: (options: any) => void;
+  xpackInfo?: any;
+}
+
+export class Plugin {
+  private readonly initializerContext: PluginInitializerContext;
+  private readonly log: Logger;
+  private readonly getLogger: (...scopes: string[]) => Logger;
+  private cluster?: ICustomClusterClient;
+
+  constructor(initializerContext: PluginInitializerContext) {
+    this.initializerContext = initializerContext;
+    this.log = initializerContext.logger.get(LOGGING_TAG);
+    this.getLogger = (...scopes: string[]) => initializerContext.logger.get(LOGGING_TAG, ...scopes);
+  }
+
+  async setup(core: CoreSetup, plugins: PluginsSetup) {
+    const [config, legacyConfig] = await combineLatest([
+      this.initializerContext.config.create<MonitoringConfig>(),
+      this.initializerContext.config.legacy.globalConfig$,
+    ])
+      .pipe(first())
+      .toPromise();
+
+    // Monitoring creates and maintains a connection to a potentially
+    // separate ES cluster - create this first
+    const elasticsearchConfig = new ElasticsearchConfig(
+      config.ui.elasticsearch as ElasticsearchConfigType
+    );
+    const cluster = (this.cluster = await instantiateClient({
```

Comment:
1. don't think `await` is required
2. do you still need `once` wrapper in `instantiateClient`?

Reply:
Good points, I'll fix